### PR TITLE
enhancements to console output

### PR DESCRIPTION
**Pull request by maboa**
This should add support to `console.log` for:

`ArrayBuffer`, `DataView`, `SharedArrayBuffer`, `Int8Array`, `Int16Array`, `Int32Array`, `Uint8Array`, `Uint16Array`, `Uint32Array`, `Uint8ClampedArray`, `Float32Array`, `Float64Array` and `Symbol`.

Notes
------

- For now I've used the Chrome/Safari style because it was slightly simpler :)
- Arrays of arrays etc support seem to work *reasonably* well, can very likely be improved.
- I have multiple `return`s in one function, this may or may not be more readable than the alternative.
- I used a regexp to match the names of object types since there are many, I hope it's still clear and readable.
- Hopefully this also fixes issue #292 

Testing
-------

I wrote a test script : 
```
console.log(42); // 42
console.log("string"); // "string"
console.log(new Array(1,2,3,4)); // [1, 2, 3, 4]
console.log(new Array("a","b","c","d")); // ["a", "b", "c", "d"]
console.log(new ArrayBuffer()); // ArrayBuffer(0) {}
console.log(new ArrayBuffer(4)); // ArrayBuffer(4) {}
console.log(new DataView(new ArrayBuffer())); // DataView(0) {}
console.log(new DataView(new ArrayBuffer(4))); // DataView(4) {}
console.log(new SharedArrayBuffer()); // SharedArrayBuffer(0) {}
console.log(new SharedArrayBuffer(4)); // SharedArrayBuffer(4) {}
console.log(new Int8Array()); // Int8Array []
console.log(new Int8Array(4)); // Int8Array(4) [0,0,0,0]
console.log(new Int16Array()); // Int16Array []
console.log(new Int16Array(4)); // Int16Array(4) [0,0,0,0]
console.log(new Int32Array()); // Int32Array []
console.log(new Int32Array(4)); // Int32Array(4) [0,0,0,0]
console.log(new Uint8Array()); // Uint8Array []
console.log(new Uint8Array(4)); // Uint8Array(4) [0,0,0,0]
console.log(new Uint16Array()); // Uint16Array []
console.log(new Uint16Array(4)); // Uint16Array(4) [0,0,0,0]
console.log(new Uint32Array()); // Uint32Array []
console.log(new Uint32Array(4)); // Uint32Array(4) [0,0,0,0]
console.log(new Uint8ClampedArray()); // Uint8ClampedArray []
console.log(new Uint8ClampedArray(4)); // Uint8ClampedArray(4) [0,0,0,0]
console.log(new Float32Array()); // Float32Array []
console.log(new Float32Array(4)); // Float32Array(4) [0,0,0,0]
console.log(new Float64Array()); // Float64Array []
console.log(new Float64Array(4)); // Float64Array(4) [0,0,0,0]
console.log(new Array(new Int8Array(), new Int16Array())); // [Int8Array [], Int16Array []]
console.log(new Array(new Int8Array(new Float64Array(4)), new Int16Array(new Float64Array(4)))); // [Int8Array(4) [0,0,0,0], Int16Array(4) [0,0,0,0]]
console.log(new Int8Array(1,1,1,1)); // Int8Array(1) [0]
console.log(Symbol('a')); // Symbol(a)
console.log(Symbol(42)); // Symbol(42); 
```

I would ❤️  to automate testing. I've been looking at Mocha.

### CODE REVIEW

File: js/editor-libs/console.js [hunk 1]
```diff
@@ -46,11 +46,46 @@
...
+     * Uint8Array, Uint16Array, Uint32Array,
+     * Uint8ClampedArray, Float32Array, Float64Array
+     * @param {any} input - The output to log.
+     * @returns Formatted output as a string.
+     */
+    function formatObject(input) {
+        var objectName = input.constructor.name;
+
+        // Re Regexp below - ^ (match at the beginning of input) and $ (match at the end of input)
+
```
Comment (schalkneethling) [inline js/editor-libs/console.js]: First pass Nit: Extra empty line

File: js/editor-libs/console.js [hunk 2]
```diff
@@ -46,11 +46,46 @@
...
+            var arrayLength = input.length;
+
+            if (arrayLength > 0) {
+                return objectName + '(' + arrayLength + ')' + ' [' + input + ']';
+            } else {
+                return objectName + ' []';
+            }
+        }
+
+        if (objectName === "Symbol") {
```
Comment (schalkneethling) [inline js/editor-libs/console.js]: Nit: Use single quotes

File: js/editor-libs/console.js [hunk 3]
```diff
@@ -46,11 +46,46 @@
...
+        return input;
+    }
+
+    /**
      * Formats output to indicate its type:
      * - quotes around strings
      * - square brackets around arrays
      * (also copes with arrays of arrays)
-     * does NOT detect Int32Array etc
+     * will cope with Int32Array etc also within arrays
```
Comment (schalkneethling) [inline js/editor-libs/console.js]: Super Nit: will cope with Int32Array etc also within arrays > `will cope with Int32Array etc. also within arrays` - Just a missing period after "etc"

**Comment by schalkneethling** [comment]
Thanks a lot for this @maboa - I did an initial pass and found tiny nits. Will pull down and test locally.

**Comment by schalkneethling** [comment]
A couple of observations:

The following: `console.log(new Int8Array(1,1,1,1));`
Prints the following:

Chrome: `Int8Array [0]`
Firefox: `Int8Array [ 0 ]`
Editor: `Int8Array(1) [0]`

The following: `console.log(new Array(new Int8Array(), new Int16Array()));`
Prints the following:

Chrome: `[Int8Array(0), Int16Array(0)]`
Firefox: `Array [ [], [] ]`
Editor: `[Int8Array [], Int16Array []]`

The following: `console.log(new SharedArrayBuffer(4));`

Prints the following:

Chrome: `SharedArrayBuffer {}`
Firefox: `SharedArrayBuffer { byteLength: 4 }`
Editor: `SharedArrayBuffer(4) {}`

**Comment by schalkneethling** [comment]
There are some other slight inconsistencies, and I know you focused on matching the blink/webkit style for now. This is also much, much better than what we have so, I am not going to block this from merging based on the above.

I merely wanted to highlight it, so we are aware. So, other than a few nits, this is r+w/c - Thanks @maboa

**Comment by schalkneethling** [comment]
@wbamberg See comment https://github.com/mdn/interactive-examples/pull/308#issuecomment-339293492

**Comment by schalkneethling** [comment]
@maboa Just waiting to see if @wbamberg has any strong feelings on this one way or the other ;) Then, there is also a conflict now with `console.js` that prevents this from being merged. I reckon you just need to pull and rebase against the latest master.

**Comment by wbamberg** [comment]
I'm really happy to see improvements to the console logging. I didn't review the code, but trust @schalkneethling 's review :). In general, I think following the conventions of browsers is a good thing as long as it doesn't make the code overly complex and on that basis a change like this is definitely worth making.

>  It looks like new Int8Array(1,1,1,1) just creates an array with one element, which I guess is where Int8Array(1) [0] comes from.

As far as I can see `new Int8Array(1,1,1,1)` is not valid according to: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Int8Array. I guess perhaps it's just ignoring all except the first argument and interpreting the first as the desired length, so it's creating an array of length 1, whose element is initialized to zero. Or maybe I'm missing something.

As I understand it, in this PR, the output for a typed array is: *\<type\>* (*\<length\>*)  [*\<comma-separated values\>*]. For example:

    console.log(new Int8Array([1,2,3]));
    > Int8Array(3) [1,2,3]

I think it would be better to be just *\<type\>* [*\<comma-separated values\>*]:

    > Int8Array [1,2,3]

...and this is more in line with what the browsers do. It's also more in line with the output for the equivalent Array call:

    console.log(new Array(1,2,3));
    > [1, 2, 3]

In fact it might be better to change the Array logic so it logs the type as well:

    > Array [1,2,3]

...which again is in line with what browsers do.

**Comment by maboa** [comment]
OK - that all seems reasonable. Thanks!

Assume for consistency we would output things like `console.log(new SharedArrayBuffer(4));` as: 
```
SharedArrayBuffer {}
```
rather than the current:
```
SharedArrayBuffer(4) {}
```

**Comment by maboa** [comment]
OK I *think* I have made the required changes and rebased correctly. Please let me know if anything is amiss. 

Here's what the expected output should look like now:

```
console.log(42); // 42
console.log("string"); // "string"
console.log(new Array(1,2,3,4)); // Array [1, 2, 3, 4]
console.log(new Array("a","b","c","d")); // Array ["a", "b", "c", "d"]
console.log(new ArrayBuffer()); // ArrayBuffer {}
console.log(new ArrayBuffer(4)); // ArrayBuffer {}
console.log(new DataView(new ArrayBuffer())); // DataView {}
console.log(new DataView(new ArrayBuffer(4))); // DataView {}
console.log(new SharedArrayBuffer()); // SharedArrayBuffer {}
console.log(new SharedArrayBuffer(4)); // SharedArrayBuffer {}
console.log(new Int8Array()); // Int8Array []
console.log(new Int8Array(4)); // Int8Array [0, 0, 0, 0]
console.log(new Int16Array()); // Int16Array []
console.log(new Int16Array(4)); // Int16Array [0, 0, 0, 0]
console.log(new Int32Array()); // Int32Array []
console.log(new Int32Array(4)); // Int32Array [0, 0, 0, 0]
console.log(new Uint8Array()); // Uint8Array []
console.log(new Uint8Array(4)); // Uint8Array [0, 0, 0, 0]
console.log(new Uint16Array()); // Uint16Array []
console.log(new Uint16Array(4)); // Uint16Array [0, 0, 0, 0]
console.log(new Uint32Array()); // Uint32Array []
console.log(new Uint32Array(4)); // Uint32Array [0, 0, 0, 0]
console.log(new Uint8ClampedArray()); // Uint8ClampedArray []
console.log(new Uint8ClampedArray(4)); // Uint8ClampedArray [0, 0, 0, 0]
console.log(new Float32Array()); // Float32Array []
console.log(new Float32Array(4)); // Float32Array [0, 0, 0, 0]
console.log(new Float64Array()); // Float64Array []
console.log(new Float64Array(4)); // Float64Array [0, 0, 0, 0]
console.log(new Array(new Int8Array(), new Int16Array())); // [Int8Array [], Int16Array []]
console.log(new Array(new Int8Array(new Float64Array(4)), new Int16Array(new Float64Array(4)))); // [Int8Array [0, 0, 0, 0], Int16Array [0, 0, 0, 0]]
console.log(new Int8Array(1,1,1,1)); // Int8Array [0]
console.log(Symbol('a')); // Symbol(a)
console.log(Symbol(42)); // Symbol(42); 
```

**Comment by schalkneethling** [comment]
@maboa Please squash your commits and resolve the conflict in `console.js` - Thx!

**Comment by maboa** [comment]
@schalkneethling  - thanks for merging :) 

My git fu could be better, so not entirely sure what happened, but the merge seems to have overwritten some of the recent changes I made.

If you run the test script above (from https://github.com/mdn/interactive-examples/pull/308#issuecomment-340729290) you should see the differences in the output compared to the `// expected output:` added in the comments.

To test, I made a new branch (https://github.com/maboa/interactive-examples/tree/console-type-enhancements-check) from this commit https://github.com/mdn/interactive-examples/commit/0e4fa92c2283ae01

Let me know if there's anything I can do to help sort out. 

ccing @wbamberg.

**Comment by schalkneethling** [comment]
@maboa I believe this one resolved the missing bits:
https://github.com/mdn/interactive-examples/pull/320

**Comment by maboa** [comment]
@schalkneethling that does indeed seem to be working now! Thanks!

**Pull request merged**